value_test.go

```diff
@@ -551,35 +551,80 @@ func TestValue_IsBoolean(t *testing.T) {
 }
 
 func TestValue_IsEqual(t *testing.T) {
-	reporter := newMockReporter(t)
-
-	data1 := map[string]interface{}{"foo": "bar"}
-	data2 := "baz"
-
-	NewValue(reporter, data1).IsEqual(data1).chain.assert(t, success)
-	NewValue(reporter, data2).IsEqual(data2).chain.assert(t, success)
-
-	NewValue(reporter, data1).NotEqual(data1).chain.assert(t, failure)
-	NewValue(reporter, data2).NotEqual(data2).chain.assert(t, failure)
-
-	NewValue(reporter, data1).IsEqual(data2).chain.assert(t, failure)
-	NewValue(reporter, data2).IsEqual(data1).chain.assert(t, failure)
-
-	NewValue(reporter, data1).NotEqual(data2).chain.assert(t, success)
-	NewValue(reporter, data2).NotEqual(data1).chain.assert(t, success)
+	t.Run("basic", func(t *testing.T) {
+		cases := []struct {
+			name         string
+			value1       interface{}
+			value2       interface{}
+			wantEqual    chainResult
+			wantNotEqual chainResult
+		}{
+			{
+				name:         "compare equivalent values (strings)",
+				value1:       "baz",
+				value2:       "baz",
+				wantEqual:    success,
+				wantNotEqual: failure,
+			},
+			{
+				name:         "compare equivalent values (maps)",
+				value1:       map[string]interface{}{"foo": "bar"},
+				value2:       map[string]interface{}{"foo": "bar"},
+				wantEqual:    success,
+				wantNotEqual: failure,
+			},
+			{
+				name:         "compare non-equivalent values",
+				value1:       map[string]interface{}{"foo": "bar"},
+				value2:       "baz",
+				wantEqual:    failure,
+				wantNotEqual: success,
+			},
+			{
+				name:         "compare nil values",
+				value1:       nil,
+				value2:       nil,
+				wantEqual:    success,
+				wantNotEqual: failure,
+			},
+			{
+				name:         "compare nil and nil-value",
+				value1:       nil,
+				value2:       map[string]interface{}(nil),
+				wantEqual:    success,
+				wantNotEqual: failure,
+			},
+			{
+				name:         "compare nil and value",
+				value1:       nil,
+				value2:       map[string]interface{}{},
+				wantEqual:    failure,
+				wantNotEqual: success,
+			},
+			{
+				name:         "compare value and func",
+				value1:       map[string]interface{}{"foo": "bar"},
+				value2:       func() {},
+				wantEqual:    failure,
+				wantNotEqual: failure,
+			},
```

Comment:
When you pass `func() {}`, the reported failure is that the argument can not be json-mashalled.

So I suggest to extract it to separate subtest "invalid argument" (like we do in many other tests).

After doing this, it seems that  it's always true that `wantEqual ==  !wantNotEqual`, so we can remove wantNotEqual here and use `!wantEqual` instead.